list links from specified page

JQinglong opened the pull request:
InstallActionに対して、-P でページを指定して、-eでリンクURLの絞り込み条件を指定するようにしました。
現時点では抽出したurlをconsoleに出力しています。
このようにして複数取得したurlについて実際のInstallAction（その他のオプション指定に従ってコンテンツの取得）を行いたいですが、そのためには、InstallActionの実体の処理を切り出して、それを呼び出すようにする必要があるのではないかと思いましたが、既存ソースへの改変が大きいため相談させていただきました。

JQinglong committed:
list links from specified page

ryo-ma commented:
> InstallActionの実体の処理を切り出して
どちらの部分のコードの想定でしょうか？

私の認識では以下の部分`installFromURL`をループで実行できれば良いと思っていますがいかがでしょうか？
https://github.com/c-3lab/dim/blob/337de174715980d2bda7a5998d074249ac9f4b09/libs/actions.ts#L82-L97

また、`parsedHeaders`も利用したいため、今回実装しているpageInstalll の処理を`parsedHeaders`の宣言以降に移動させたほうがいいですね。
最終的には、今回実装のpageInstallに関しては`action_helper/installer.ts`の方にまとめてしまいたいと思ってます。
でも、まずは一旦は動く様にしましょうか。

JQinglong committed:
implement page install

JQinglong committed:
create serial number dir under name option dir

JQinglong committed:
create serial number dir under name option dir

ryo-ma commented:
下部のCIで落ちてしまっているので`deno fmt`コマンドによってフォーマットしていただければと思います。

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
+          if (re.test(href)) {
+            idx += 1;
+            let dataName = "";
+            if (options.name !== undefined) {
```
`name`の仕様ですが必須にしようかと思います。
必ず `-n` をつけてもらうようにしましょう。
その場合は dataNameは以下の処理で十分です。

```
dataName = `${options.name}_${idx}`;
```

ryo-ma replied:
`installFromURL`に渡す`name`はディレクトリ名で、
ファイル名は`downloader.ts`でURL pathお末尾が自動でファイル名となるようになっております。
https://github.com/c-3lab/dim/blob/31a9ae781b357973b8ada6c13d2c8bd9e32a6692/libs/downloader.ts#L13

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
+                Colors.red("Failed to install."),
+                Colors.red(error.message)
+              );
+              Deno.exit(1);
```
逆にここは`Deno.exit(1)`してしまうと以下のようなページで途中で処理が止まってしまうのは不便なような気がします。エラーは出しつつ処理は進めてもいい気がします。

例：
https://xxxxx/xx1.csv <- status 200が返ってくる
https://xxxxx/xx2.csv <- status 404が返ってくる
https://xxxxx/xx3.csv <- status 200 が返ってくる(ここのデータが取得できない)

ryo-ma replied:
後エラーメッセージには対象のURLも出してあげるとユーザフレンドリーかと思います。

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
 
     const parsedHeaders: Record<string, string> = parseHeader(options.headers);
+
+    if (options.pageInstall !== undefined) {
```
また、URLを指定してしまった場合も処理が走ってしまうので その`if`の分岐として`else if`でこの処理を追加する方がよさそうですね。
以下のifです。
https://github.com/c-3lab/dim/blob/31a9ae781b357973b8ada6c13d2c8bd9e32a6692/libs/actions.ts#L84

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
     }
```
こちらの条件に `pageInstall`も含めて欲しいです。
https://github.com/c-3lab/dim/blob/31a9ae781b357973b8ada6c13d2c8bd9e32a6692/libs/actions.ts#L33

以下の組み合わせでバリデーションする必要がありそうです・
url && options.file でラーメッセージ
options.pageInstall && url でエラーメッセージ
options.file && options.pageInstall でエラーメッセージ

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
+        const linklist = document.getElementsByTagName("a");
+        let idx = 0;
+        for (const link of linklist) {
+          const re = new RegExp(options.expression, "ig");
```
ここの正規表現フラグで `i`を含めた理由は何かありますか？

JQinglong replied:
正規表現の指定を可能としつつ、基本的な指定方法は、リンク先のファイル名を部分一致で抽出することを想定しています。その場合、大文字小文字のブレは許容したいと思いました。
もっとも、iフラグを指定しなくても、大文字小文字両方を意識した正規表現を書けば良いので、必須ではないと思います。

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
+    if (options.pageInstall !== undefined) {
+      try {
+        const getResult = await fetch(options.pageInstall);
+        if (!getResult.ok) throw new Error("Fetch response error");
```
逆にここはDeno.exit(1)してしまうと以下のようなページで途中で処理が止まってしまうのは不便なような気がします。エラーは出しつつ処理は進めてもいい気がします。

例：
https://xxxxx/xx1.csv <- status 200が返ってくる
https://xxxxx/xx2.csv <- status 404が返ってくる
https://xxxxx/xx3.csv <- status 200 が返ってくる(ここのデータが取得できない)

ryo-ma replied:
後エラーメッセージには対象のURLも出してあげるとユーザフレンドリーかと思います。

ryo-ma commented on libs/actions.ts:
```diff
@@ -36,6 +38,49 @@ export class InstallAction {
+                Colors.red("Failed to install."),
+                Colors.red(error.message)
+              );
+              Deno.exit(1);
```
`Deno.exit(1)`を実行してしまうと以下のようなリンクを持ったページだと途中で処理が止まってしまうのは不便なような気がします。エラー文はconsole.log出しつつ処理は進めてもいい気がします。

例：
https://xxxxx/xx1.csv <- status 200が返ってくる
https://xxxxx/xx2.csv <- status 404が返ってくる
https://xxxxx/xx3.csv <- status 200 が返ってくる(ここのデータが取得できない)

ryo-ma replied:
後エラーメッセージには対象のURLも出してあげるとユーザフレンドリーかと思います。

Commit:
code format

JQinglong committed:
change conditional branch

JQinglong committed:
Merge branch 'c-3lab:main' into feature/page-install

JQinglong committed:
add validation

JQinglong committed:
deno fmt

JQinglong committed:
deno fmt

JQinglong committed:
remove type error

JQinglong committed:
add pageInstall test

ryo-ma commented:
CIがまたもや`deno fmt`で落ちてますね。

ryo-ma commented:
一通り修正が完了したら今回の処理部分の try の中の処理を 他のインストール処理と同様`installFromPage()`のような感じで `libs/action_helper/installer.ts`に処理をまとめたいですね。

JQinglong committed:
improve message

JQinglong committed:
deno fmt

JQinglong committed:
deno fmt

ryo-ma commented on libs/actions.ts:
```diff
@@ -57,18 +77,55 @@ export class InstallAction {
+              parsedHeaders,
+            ).catch((error) => {
+              console.log("Failed to pageInstall.");
+              console.log("target:" + href);
```
ここの三行まだ以下の記載になってませんね。
`console.error(Colors.red(""))`

JQinglong committed:
move pageinstall procedure to action_helper

JQinglong committed:
add pageinstall test

ryo-ma commented:
Github Action(CI)のlintでひっかかっちゃいましたね。 `deno lint` でチェックすると出てきます。

JQinglong committed:
check lint

JQinglong committed:
add pageinstall error test

ryo-ma commented:
また少しフォーマットでCIが落ちてしまってますね。

Commit:
add test for error catch

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# Committer: JQinglong <konnokiyotaka@gmail.com>
#
# On branch feature/page-install
# Your branch is ahead of 'origin/feature/page-install' by 1 commit.
#   (use "git push" to publish your local commits)
#
# Changes to be committed:
#	modified:   libs/action_helper/installer.ts
#	modified:   libs/actions.ts
#	modified:   tests/libs/actions.install.test.ts
#	new file:   tests/test_data/test-page-empty.html
#	new file:   tests/test_data/test-page-invalid.html
#

JQinglong committed:
Merge branch 'c-3lab:main' into feature/page-install

ryo-ma commented:
@JQinglong 
あと以下の行だけですね。ここの分岐を通せたりしますか？それができて、あらためて動作確認したらマージしたいと思います。

https://app.codecov.io/gh/c-3lab/dim/pull/107?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=c-3lab#4663a8fc79877a2f7b91c1d0c139accd-R322

Commit:
delete unnecessary lines

JQinglong commented:
@ryo-ma 
未実行になっている行ですが、何が問題でこれが必要だったのかわからなくなってしまったので、一旦外しました。
また前段階で何かのチェックにかかってしまうかもしれませんが、すみません・・・

ryo-ma commented:
@JQinglong ありがとうございます。エラー処理周りは少し気になるところもありますが、とりあえずマージしてしまいましょうか。使っている中で問題があればまたプルリクで対応してきましょう。

JQinglong commented:
マージありがとうございました。ぜひ、自分の開発でも使って、フィードバックもしていきたいと思っています。
